# Kenya Agriculture Data Analysis using PySpark

## Data Loading

This notebook demonstrates the data loading process for the Kenya Agriculture dataset using Apache Spark.

The objectives are to:

- Create a Spark session
- Load the dataset
- Inspect the data
- Verify the schema
- Check data quality before preprocessing

In [56]:
%load_ext autoreload
%autoreload 2

In [50]:
import importlib
import scripts.load_data

importlib.reload(scripts.load_data)

<module 'scripts.load_data' from 'c:\\Users\\ADMIN\\OneDrive\\Desktop\\OUK Msc\\1.2\\Data Mining and Big Data\\Big-Data-Analytics-with-Hadoop-and-Spark\\scripts\\load_data.py'>

In [30]:
import os
import sys

sys.path.append(os.path.abspath(".."))

from scripts.load_data import (
    create_spark_session,
    load_dataset,
    preview_data,
    dataset_shape,
    dataset_schema,
    list_columns,
    dataset_summary,
    check_missing_values,
)

In [31]:
spark = create_spark_session()

In [32]:
# Load the dataset 

df = load_dataset(
    spark,
    "../data/kenya_agriculture_raw.csv"
)

Dataset loaded successfully.


## Preview Dataset

The first five rows are displayed to verify that the dataset has been loaded correctly and to understand its overall structure.

In [33]:
preview_data(df)

+----+---------+----------+---------+---------------+-----+---------+---------------+--------------------+------------+--------------+---------+----+----+-----+----+----+
|Iso3|StartDate|   EndDate|Area Code|Area Code (M49)| Area|Item Code|Item Code (CPC)|                Item|Element Code|       Element|Year Code|Year|Unit|Value|Flag|Note|
+----+---------+----------+---------+---------------+-----+---------+---------------+--------------------+------------+--------------+---------+----+----+-----+----+----+
| KEN| 1/1/1991|12/31/1991|      114|           '404|Kenya|      809|      '01929.07|Abaca, manila hem...|        5312|Area harvested|     1991|1991|  ha| 25.0|   E|NULL|
| KEN| 1/1/1992|12/31/1992|      114|           '404|Kenya|      809|      '01929.07|Abaca, manila hem...|        5312|Area harvested|     1992|1992|  ha| 26.0|   E|NULL|
| KEN| 1/1/1993|12/31/1993|      114|           '404|Kenya|      809|      '01929.07|Abaca, manila hem...|        5312|Area harvested|     1993|1

## Dataset Dimensions

The number of rows and columns provides an overview of the dataset size.

In [34]:
dataset_shape(df)

(24233, 17)

## Dataset Schema

The schema shows the name and data type of each column. Spark automatically inferred the appropriate data types during loading.

In [35]:
dataset_schema(df)

root
 |-- Iso3: string (nullable = true)
 |-- StartDate: string (nullable = true)
 |-- EndDate: string (nullable = true)
 |-- Area Code: integer (nullable = true)
 |-- Area Code (M49): string (nullable = true)
 |-- Area: string (nullable = true)
 |-- Item Code: integer (nullable = true)
 |-- Item Code (CPC): string (nullable = true)
 |-- Item: string (nullable = true)
 |-- Element Code: integer (nullable = true)
 |-- Element: string (nullable = true)
 |-- Year Code: integer (nullable = true)
 |-- Year: integer (nullable = true)
 |-- Unit: string (nullable = true)
 |-- Value: double (nullable = true)
 |-- Flag: string (nullable = true)
 |-- Note: string (nullable = true)



## Column Names

Listing the column names helps verify that all expected variables have been loaded.

In [36]:
print(list_columns(df))

['Iso3', 'StartDate', 'EndDate', 'Area Code', 'Area Code (M49)', 'Area', 'Item Code', 'Item Code (CPC)', 'Item', 'Element Code', 'Element', 'Year Code', 'Year', 'Unit', 'Value', 'Flag', 'Note']


## Summary Statistics

Descriptive statistics provide an overview of the numeric variables, including the count, mean, standard deviation, minimum, and maximum values.

In [37]:
dataset_summary(df)

+-------+-----+---------+----------+---------+---------------+-----+------------------+---------------+--------------------+-----------------+--------------------+------------------+------------------+-------+------------------+-----+-----------------+
|summary| Iso3|StartDate|   EndDate|Area Code|Area Code (M49)| Area|         Item Code|Item Code (CPC)|                Item|     Element Code|             Element|         Year Code|              Year|   Unit|             Value| Flag|             Note|
+-------+-----+---------+----------+---------+---------------+-----+------------------+---------------+--------------------+-----------------+--------------------+------------------+------------------+-------+------------------+-----+-----------------+
|  count|24233|    24233|     24233|    24233|          24233|24233|             24233|          24233|               24233|            24233|               24233|             24233|             24233|  24233|             23502|24233|       

## Missing Values

The dataset is checked for missing values to ensure completeness before any preprocessing or analysis.

In [38]:
check_missing_values(df)

+----+---------+-------+---------+---------------+----+---------+---------------+----+------------+-------+---------+----+----+-----+----+-----+
|Iso3|StartDate|EndDate|Area Code|Area Code (M49)|Area|Item Code|Item Code (CPC)|Item|Element Code|Element|Year Code|Year|Unit|Value|Flag| Note|
+----+---------+-------+---------+---------------+----+---------+---------------+----+------------+-------+---------+----+----+-----+----+-----+
|   0|        0|      0|        0|              0|   0|        0|              0|   0|           0|      0|        0|   0|   0|  731|   0|23708|
+----+---------+-------+---------+---------------+----+---------+---------------+----+------------+-------+---------+----+----+-----+----+-----+



# Data Cleaning

This section evaluates and improves the quality of the dataset by checking for duplicate records, missing values, standardized column names, and invalid numeric values.

In [53]:
import importlib
import scripts.clean_data

importlib.reload(scripts.clean_data)

<module 'scripts.clean_data' from 'c:\\Users\\ADMIN\\OneDrive\\Desktop\\OUK Msc\\1.2\\Data Mining and Big Data\\Big-Data-Analytics-with-Hadoop-and-Spark\\scripts\\clean_data.py'>

In [54]:
from scripts.clean_data import (
    check_duplicates,
    remove_duplicates,
    check_missing_values,
    remove_missing_values,
    standardize_column_names,
    validate_numeric_columns,
    data_quality_report,
)

In [40]:
# Number of duplicate rows
duplicates = check_duplicates(df)
print(f"Duplicate Rows: {duplicates}")

Duplicate Rows: 0


In [41]:
# Missing values
check_missing_values(df)

+----+---------+-------+---------+---------------+----+---------+---------------+----+------------+-------+---------+----+----+-----+----+-----+
|Iso3|StartDate|EndDate|Area Code|Area Code (M49)|Area|Item Code|Item Code (CPC)|Item|Element Code|Element|Year Code|Year|Unit|Value|Flag| Note|
+----+---------+-------+---------+---------------+----+---------+---------------+----+------------+-------+---------+----+----+-----+----+-----+
|   0|        0|      0|        0|              0|   0|        0|              0|   0|           0|      0|        0|   0|   0|  731|   0|23708|
+----+---------+-------+---------+---------------+----+---------+---------------+----+------------+-------+---------+----+----+-----+----+-----+



In [42]:
df.filter(col("Value").isNull()).show(10, truncate=False)

+----+---------+----------+---------+---------------+-----+---------+---------------+-------------------------------------------------------------------------+------------+--------------+---------+----+----+-----+----+----+
|Iso3|StartDate|EndDate   |Area Code|Area Code (M49)|Area |Item Code|Item Code (CPC)|Item                                                                     |Element Code|Element       |Year Code|Year|Unit|Value|Flag|Note|
+----+---------+----------+---------+---------------+-----+---------+---------------+-------------------------------------------------------------------------+------------+--------------+---------+----+----+-----+----+----+
|KEN |1/1/1961 |12/31/1961|114      |'404           |Kenya|711      |'01654         |Anise, badian, coriander, cumin, caraway, fennel and juniper berries, raw|5312        |Area harvested|1961     |1961|ha  |NULL |M   |NULL|
|KEN |1/1/1962 |12/31/1962|114      |'404           |Kenya|711      |'01654         |Anise, badian, cori

### Missing Values Summary

The missing value assessment indicates that most variables are complete.

Two columns contain missing values:

- **Value** contains 731 missing observations. Since this variable represents the agricultural measurement, these records require further investigation before analysis.
- **Note** contains 23,708 missing values. This column stores optional annotations and therefore its missing values are expected and do not necessarily affect data quality.

Overall, the dataset is largely complete, with missing values concentrated in a small number of fields.

In [55]:
from scripts.clean_data import remove_missing_values
clean_df = remove_missing_values(df)

print("Rows before cleaning:", df.count())
print("Rows after cleaning:", clean_df.count())

Rows before cleaning: 24233
Rows after cleaning: 23502


### Handling Missing Values

The missing value assessment showed that the `Value` column contains 731 missing observations. Inspection revealed that these records are marked with the flag `M`, indicating that the measurements were unavailable rather than equal to zero.

To preserve the integrity of the dataset, missing values were not imputed. Instead, records with missing `Value` entries were excluded from the cleaned dataset used for subsequent transformations and analysis.

The `Note` column contains many missing values because it stores optional annotations. These missing values do not affect the analytical variables and were therefore retained.

In [63]:
# Standardize column names

from scripts.clean_data import standardize_column_names
clean_df = standardize_column_names(clean_df)
print(clean_df.columns)

['iso3', 'startdate', 'enddate', 'area_code', 'area_code_m49', 'area', 'item_code', 'item_code_cpc', 'item', 'element_code', 'element', 'year_code', 'year', 'unit', 'value', 'flag', 'note']


In [66]:
clean_df.filter(col("value") < 0).count()

0

In [62]:
# Validate numeric values
invalid = validate_numeric_columns(clean_df)
invalid.show(5)

Negative values found: 0
+----+---------+-------+---------+---------------+----+---------+---------------+----+------------+-------+---------+----+----+-----+----+----+
|Iso3|StartDate|EndDate|Area Code|Area Code (M49)|Area|Item Code|Item Code (CPC)|Item|Element Code|Element|Year Code|Year|Unit|Value|Flag|Note|
+----+---------+-------+---------+---------------+----+---------+---------------+----+------------+-------+---------+----+----+-----+----+----+
+----+---------+-------+---------+---------------+----+---------+---------------+----+------------+-------+---------+----+----+-----+----+----+



In [65]:
# Data quality report
data_quality_report(clean_df)

Rows: 23502
Columns: 17
Duplicates: 0
Missing Values
+----+---------+-------+---------+-------------+----+---------+-------------+----+------------+-------+---------+----+----+-----+----+-----+
|iso3|startdate|enddate|area_code|area_code_m49|area|item_code|item_code_cpc|item|element_code|element|year_code|year|unit|value|flag| note|
+----+---------+-------+---------+-------------+----+---------+-------------+----+------------+-------+---------+----+----+-----+----+-----+
|   0|        0|      0|        0|            0|   0|        0|            0|   0|           0|      0|        0|   0|   0|    0|   0|22977|
+----+---------+-------+---------+-------------+----+---------+-------------+----+------------+-------+---------+----+----+-----+----+-----+



## Data Cleaning Summary

The dataset was successfully cleaned and validated before further analysis.

The cleaning process included:

- Checking for duplicate records and confirming that none remained.
- Removing records with missing values in the `value` column, as these represent unavailable agricultural measurements.
- Standardizing column names by converting them to lowercase and replacing spaces with underscores.
- Validating the dataset to ensure all analytical variables were complete.

Following the cleaning process, the dataset contains **23,502 records** and **17 variables**. The only remaining missing values occur in the `note` column, which contains optional metadata and does not affect the analytical variables used in this project.